src/main/java/graphics/renderer/Renderer.java

```diff
@@ -47,168 +47,171 @@
-            rebuffer();
-            finish();
-        }
-
-        for (RenderBatch batch : batches) {
-            batch.bind();
-            glDrawElements(batch.primitive.openglPrimitive, batch.getVertexCount(), GL_UNSIGNED_INT, 0);
-            batch.unbind();
-        }
-        shader.detach();
-        Framebuffer.unbind();
-    }
-
-    /**
-     * Prepare for rendering. Do anything like setting background here.
-     */
-    protected abstract void prepare();
-
-    /**
-     * Delete all the Batches.
-     */
-    public void clean() {
-        batches.forEach(RenderBatch::delete);
-    }
+	/**
+	 * Texture slots to be uploaded to the shader. You don't have to upload them in your custom renderer.
+	 */
+	protected final int[] textureSlots = {0, 1, 2, 3, 4, 5, 6, 7};
+
+	/**
+	 * A list of batches
+	 */
+	protected final List<RenderBatch> batches;
+	/**
+	 * Framebuffer to which this renderer will render
+	 */
+	public Framebuffer framebuffer;
+	protected boolean noRebuffer = false;
+	/**
+	 * Shader to be used for rendering
+	 */
+	private Shader shader;
+
+	public Renderer() {
+		this.batches = new OrderPreservingList<>();
+	}
+
+	/**
+	 * Create a shader
+	 *
+	 * @return the created shader
+	 */
+	protected abstract Shader createShader();
+
+	/**
+	 * Create a framebuffer
+	 *
+	 * @return the created fbo
+	 */
+	protected abstract Framebuffer createFramebuffer();
+
+	/**
+	 * Create a new Batch with appropriate parameters
+	 *
+	 * @return a new batch
+	 */
+	protected abstract RenderBatch createBatch(int zIndex);
+
+	/**
+	 * Upload the required uniforms
+	 *
+	 * @param shader the shader
+	 */
+	protected abstract void uploadUniforms(Shader shader);
+
+	/**
+	 * Rebuffer all the data into batches
+	 */
+	protected abstract void rebuffer();
+
+	/**
+	 * Start batches. Ready for buffering data
+	 */
+	private void start() {
+		for (RenderBatch batch : batches) {
+			batch.start();
+		}
+	}
+
+	/**
+	 * Finish Setting data for all batches. Upload to gpu ready for rendering
+	 */
+	private void finish() {
+		for (RenderBatch batch : batches) {
+			batch.finish();
+		}
+	}
+
+	/**
+	 * Get the batch in which the current data can be submitted
+	 * Has to be called PER PRIMITIVE SUBMISSION
+	 *
+	 * @param texture
+	 */
+	public RenderBatch getAvailableBatch(Texture texture, int reqdZ) {
+		for (RenderBatch batch : batches) {
+			if (!batch.isFull && batch.zIndex() == reqdZ)
+				return batch;
+
+			if (batch.isFull_Textures)
+				if (batch.hasTexture(texture) && batch.zIndex() == reqdZ)
+					return batch;
+		}
+
+		// All batches full
+		RenderBatch batch = createBatch(reqdZ);
+		batch.init();
+		batch.start();
+		batches.add(batch);
+		return batch;
+	}
+
+	/**
+	 * Add a gameObject to the renderer, and if it contains a component that affects rendering, like a sprite or light, those are added to a batch.
+	 *
+	 * @param gameObject the GameObject with renderable components
+	 */
+	public void add(GameObject gameObject) {

```

Comment:
Do we need this method - although not every Renderer is gameobject related?

Reply:
same goes for remove some lines below this

Reply:
It seems like it would make more sense for this to be an abstract method with type Object or something along those lines.

Reply:
Or maybe even a generic, I mean, we really dont use much of them and this case sounds perfect to use them